kasa/klapprotocol.py

```diff
@@ -0,0 +1,500 @@
+
+        if response_status != 200:
+            raise AuthenticationException(
+                f"Device {self.host} responded with {response_status} to handshake1 {self.discovery_data}"
+            )
+        self.handle_cookies(session, url)
+
+        remote_seed = response_data[0:16]
+        server_hash = response_data[16:]
+
+        cookie = self.jar.filter_cookies(url).get(self.TP_SESSION_COOKIE_NAME)
+        _LOGGER.debug(
+            f"Handshake1 posted at {datetime.datetime.now()}.  Host is {self.host}, Session cookie is {cookie}, Response status is {response_status}, Request was {self.local_auth_hash.hex()}"
+        )
+
+        _LOGGER.debug(
+            "Server remote_seed is: %s, server hash is: %s",
+            remote_seed.hex(),
+            server_hash.hex(),
+        )
+
+        local_seed_auth_hash = self._sha256(self._local_seed + self.local_auth_hash)
+
+        # Check the response from the device
+        if local_seed_auth_hash == server_hash:
+            _LOGGER.debug("handshake1 hashes match")
+            return remote_seed, self.local_auth_hash
+        else:
+            _LOGGER.debug(
+                "Expected %s got %s in handshake1.  Checking if blank auth is a match",
+                local_seed_auth_hash.hex(),
+                server_hash.hex(),
+            )
+
+            blank_auth = Credentials(username="", password="")
+            blank_auth_hash = self.generate_auth_hash(blank_auth)
+            blank_seed_auth_hash = self._sha256(self._local_seed + blank_auth_hash)
+            if blank_seed_auth_hash == server_hash:
+                _LOGGER.debug(
+                    "Server response doesn't match our expected hash on ip %s but an authentication with blank credentials matched",
+                    self.host,
+                )
+                return remote_seed, blank_auth_hash
+            else:
+                kasa_setup_auth = Credentials(
+                    username=self.KASA_SETUP_EMAIL, password=self.KASA_SETUP_PASSWORD
+                )
+                kasa_setup_auth_hash = self.generate_auth_hash(kasa_setup_auth)
+                kasa_setup_seed_auth_hash = self._sha256(
+                    self._local_seed + kasa_setup_auth_hash
+                )
+                if kasa_setup_seed_auth_hash == server_hash:
+                    self.local_auth_hash = kasa_setup_auth_hash
+                    _LOGGER.debug(
+                        "Server response doesn't match our expected hash on ip %s but an authentication with kasa setup credentials matched",
+                        self.host,
+                    )
+                    return remote_seed, kasa_setup_auth_hash
+                else:
+                    self.authentication_failed = True
+                    msg = f"Server response doesn't match our challenge on ip {self.host} {self.discovery_data}"
+                    _LOGGER.debug(msg)
```

Comment:
Logging should be lazy as to avoid formatting strings that will be thrown away most of the time

https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/logging-not-lazy.html

Reply:
The line after this raises an exception with the msg so this should be ok `raise AuthenticationException(msg)`

Reply:
Yup, this is fine here.